add toggle to switch between ply file and original file in 3d viewer

ErikAmerico opened the pull request:
Added a toggle button to switch between viewing the ply file and original file.

ErikAmerico committed:
add toggle to switch between ply file and original file in 3d viewer

ErikAmerico commented on src/components/AnalysisPage/ThicknessAnalysisBox/ThickAnalysisBox.jsx:
```diff
@@ -76,6 +77,7 @@ const ThickAnalysisBox = ({
         setFileFor3dModel(file);
+        setShowToggle(true);
       } else {
         if (listOfThicknessData.find((data) => data.id === data.id)) {
```
if downloadPlyFilePlaceIn3dViewer fails, will setShowToggle(true) still run? or will this whole if statement just cancel?

I wasn't able to pass setShowToggle to downloadPlyFile to handle this in that function

rknizzle replied:
If downloadPlyFilePlaceIn3dViewer throws an error then setShowToggle(true) will not run. Throwing an error stops the code executing. When an error is thrown it 'bubbles up' until it goes up into a try in which case it will stop bubbling up and go into the catch block.

Heres an example of bubbling up:

```js
function top() {
  try {
    doThing()
  } catch(err) {
    console.log('catching the error, maybe there would be a UI popup or something')
  }
}

function lower() {
  lowest()
  console.log('2')
  otherThing()
}

function lowest() {
  console.log('1')
  throw new Error('a bad thing happened')
}


top()

```

This program logs '1' and then 'catching the error, maybe there would be a UI popup or something'.

It does not log '2'.

So  this shows an error bubbling up. As functions call other functions it creates new lower layers and so when one of those lower layers throws an error, it goes back up to the higher layers either until it goes into a try in which case it will stop bubbling up and go into the catch block or iif it never goes into a try block itll just bubble up until it gets to the top which would be probably the React component

ErikAmerico commented:
Oh and we don't need to add this toggle button, I just know you mentioned having one. 

Up to you!

rknizzle commented:
@ErikAmerico I did want a toggle but now with how you made the UI easily switch back to grey when you delete a value in the threshold input I don't t hink its as necessary but it might still be good. What do you think about the toggle UI?

But I do think that I definitely don't want to release the toggle feature until we figure out how to change the color of the model without reloading the whole scene so it doesn't cause the black screen flash and the resetting of the model position. If we add the toggle I'd like it to just change the color and not reset the position of the model

ErikAmerico commented:
I don't think the toggle is completely useless.

If someone has a 1 threshold value result saved and they are currently viewing a 1.7. When they wanted to switch back to the gray model, 

they wild have to delete the .7, which would briefly show the ply file for the 1 value then they can delete the 1 to get the gray model.

and I personally felt trapped when viewing the ply model. Now I feel like I'm free to look at either 😂

ErikAmerico commented:
I'll start getting into switching the color without resetting the scene! And keeping the same position.

ErikAmerico committed:
create seperate functions for loading and clearing models

ErikAmerico committed:
setUpThreeScene once on inital load, then use loadModel() and clearModel() to manage which file is shown in the viewer

ErikAmerico committed:
keep position of model when toggling, reset position of a newly viewed model.

ErikAmerico committed:
rename isNewOriginalFileNameFor3dModel -> isNewOriginalFileFor3dModel

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -2,14 +2,29 @@ import { useEffect, useRef, useState } from "react";
+  useEffect(() => {
+    if (fileFor3dModel === originalFileFor3dModel) {
+      setShowToggle(false);
+    }
```
Can showToggle and setShowToggle be totally managed by the ModelViewer component? Does it need to be state in the Analysis page and passed around to a bunch of other components?

ErikAmerico replied:
yes it can, and super easily. Thanks dude!

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -15,67 +15,48 @@ const ModelViewer = ({
   const canvasRef = useRef(null);
-  const [objectURL, setObjectURL] = useState(null);
-  const [fileType, setFileType] = useState(null);
+  const sceneRef = useRef(null);
```
Why is this a 'Ref' and why does it use the property current? Why isn't it just like modelViewerSceneHandler or something? Is there some reason why thats necessary? I'm not super familiar with what refs are in React

ErikAmerico replied:
I'm also not an expert in react Refs.

But here its a reference to the setUpThreeScene function call.

To the best of my knowledge

Using ref lets us load and clear new models into the scene without having to re-setup the scene every time a new model loads or clears. Where if we were using a modelViewerSceneHandler state, it would trigger a re-render on each change and we would be calling setUpThreeScene over and over again (like we were before i believe, causing that black flash).

.current is where the reference is located in sceneRef. Here is the output of `console.log(sceneRef)` 

![image](https://github.com/rknizzle/propermesh-view/assets/111093132/51c2ea6c-e6f5-4b70-86e2-5fcdaa50ca7f)

rknizzle commented on src/components/AnalysisPage/3dModelViewer/setupThreeScene.js:
```diff
@@ -138,11 +148,15 @@ const setupThreeScene = (canvas, objectURL, fileType) => {
   };
+  animate();
+
+  return { loadModel, clearModel };
```
Really clever solution here returning an object containing functions to modify the scene from the React component

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -15,67 +15,48 @@ const ModelViewer = ({
+      const objectURL = URL.createObjectURL(fileFor3dModel);
+      sceneRef.current.loadModel(objectURL, fileType);
+      URL.revokeObjectURL(objectURL);
+      sceneRef.current.clearModel();
```
Wait why are you loading and then clearing right after?

ErikAmerico replied:
hmmm really only because it just works this way. If i remove `sceneRef.current.clearModel()`, then the next time i load a model. the new model will be overlapping the previous model. 

But I think whats actually happening is that when this code runs, `sceneRef.current.clearModel()` doesn't actually have anything to clear because the model hasn't been added to sceneRef.current yet. Then the next time the useEffect runs, the new model will be in the process of loading while `sceneRef.current.clearModel()` gets hit and clears the previously loaded model.

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -17,10 +17,8 @@ const ModelViewer = ({
-    previousOriginalFileNameFor3dModel,
-    setPreviousOriginalFileNameFor3dModel,
-  ] = useState(null);
+  const [previousOriginalFileFor3dModel, setPreviousOriginalFileFor3dModel] =
```
Just a note on commit message for cbdfbceb7b2f542f73557f64b600f202f9bbc00a:
You renamed 2 variables in the commit but only mentioned one of them in the commit message. So just be careful of that cause it can cause a little confusion. You could have either split this into 2 commits or just mentioned both renames in the commit.

I use the command `git diff --staged` very often (I created a terminal shortcut `gds` for it). It shows all the changes that you have added and are going to get commited when you run `git commit`. This can be really useful for checking whats about to be committed.

Also commits contain sort of 2 parts; a title and a message. The title is just a one line thing and then the message can be as long as you want it to be. So you could have done like:

```
Refactor: Rename variables

previousOriginalFileNameFor3dModel -> previousOriginalFileFor3dModel
&
isNewOriginalFileNameFor3dModel -> isNewOriginalFileFor3dModel
```

Where the top line is the title and the rest is the message. 

The way that you commit is just kind of only a title that extends to the next line if it gets too long.

Heres an example of a commit that has a title and a message: https://github.com/rknizzle/propermesh/commit/77ae192b77760a33c3d7f64f1fd9563f4ba844d8

The title vs message thing isn't a big deal though so you don't need to worry about changing the way you commit right now. I just wanted to introduce you to the concept

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -16,18 +16,34 @@ const ModelViewer = ({
+      originalFileFor3dModel !== previousOriginalFileNameFor3dModel;
+
+    const viewingNewPart =
+      isNewFileFor3dModel && isNewOriginalFileNameFor3dModel;
```
THis is starting to get a little confusing with all the current vs original vs previous stuff for everything. Im not quite sure how to make it better though. Maybe just refactoring to make some small well named functions or something.

Or maybe theres a way to split it apart so that instead of all the loading logic being here we could have like a function to load in a new original file and then a separate function to load the ply visualization file into the viewer.

Not sure just thinking out loud, we can keep an eye on this as we continue developing the viewer

ErikAmerico replied:
In one of my latest commits, I was able to completely remove all of this! 

no more confusion

rknizzle commented on src/components/AnalysisPage/3dModelViewer/setupThreeScene.js:
```diff
@@ -47,86 +47,96 @@ const setupThreeScene = (canvas, objectURL, fileType) => {
+    } else if (fileType === "ply") {
+      loader = new PLYLoader();
+    } else {
+      throw new Error(`Unsupported file type`);
```
The toggle works a lot better now but I think theres one more optimization we can make when switching between the grey model and the thickness visualization model.  Once we load in a ply file, instead of calling the loader every time to load back in the original file to get the grey colored view, we can just change the color of the model using threejs materials. So instead of loading in the original part file every time to get the grey view we can just change the material on the mesh to show grey.

This will solve the little pause between switching out the models too. Currently if you load in a really big file it can take 4-5 seconds to switch between the views when you press the toggle because it needs to load the whole part into the loader each time.

You've already mentioned wanting to fix the little flash of the part when switching the toggle so I think this might be a good way to go about it

ErikAmerico replied:
This should be good now too!

rknizzle reviewed (commented):
Nice solution! I left a couple little changes and some questions here and then I'll merge this in when you address those. I also gave my idea for an optimization to fix the pause between showing the model when you toggle the switch. This works well how it is though so I'm also down to merge this in before you fully fix the pause between models. As soon as you answer my couple little questions and possibly make some updates for them I'm good to merge this in

ErikAmerico committed:
manage showToggle and setShowtoggle directly in ModelViewer

remove showToggle and setShowToggle from thickanalysisbox

remove showToggle and setShowToggle from start button

ErikAmerico committed:
Call clearModel() in loadModel()

- remove sceneRef.current.clearModel(); from
useEffect hooks in 3dModelViewer.jsx

Calling sceneRef.current.clearModel();
after loadModel() in the useEffect hook
would clear the previous model. But only because
the model that was being loaded in the useEffect
hook hadn't loaded yet. I think if the loading
process finished quickly for some reason,
the model that we wanted to see would end up
being cleared.

ErikAmerico committed:
Add comments

* Just explaining why I think
URL.revokeObjectURL(objectURL); is placed correctly.

* unlike how sceneRef.current.clearModel() was used
in before my previous commit, which caused
timing issues.

ErikAmerico committed:
Toggle between mesh colors instead of files

- made a toggleMesh function in setUpThreeScene.js
  and used it in 3dModelViewer.jsx

- now we do not load a new file when we want to
  change the color of the mesh, we just change the
  color of the mesh that is already loaded.

- the black flash when using the toggle button is gone,
  but it is still there when you mess with the
  threshold value. So i'll check that out next.

- Tomorow I am going to go through and remove any
  pieces of state that are being passed down into
  components that are not necessary anymore after
  this change. `originalFileNameFor3dModel` is
  an obvious one that can be removed from
    `3dModelViewer.jsx`.

rknizzle commented on src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:
```diff
@@ -49,6 +48,9 @@ const ModelViewer = ({
     if (fileFor3dModel === originalFileFor3dModel) {
       setShowToggle(false);
     }
+    if (fileFor3dModel !== originalFileFor3dModel) {
```
I think just using an else here would be simpler and clearer

rknizzle commented on src/components/AnalysisPage/3dModelViewer/setupThreeScene.js:
```diff
@@ -81,6 +82,15 @@ const setupThreeScene = (canvas) => {
       const mesh = new THREE.Mesh(geometry, material);
 
+      // this if statement has to be seperate from similar statement
+      // on line 76 because the materials have to be defined before the mesh
```
Try to avoid specifying particular line numbers in code because if you add a new feature in your next PR that pushes the code down, the correct code won't line up with this comment anymore. Try to just use descriptions instead

rknizzle commented on src/components/AnalysisPage/3dModelViewer/setupThreeScene.js:
```diff
@@ -128,6 +138,16 @@ const setupThreeScene = (canvas) => {
     });
   };
 
+  const toggleMesh = (fileType) => {
```
I think this function could use a bit more detailed of a name to include what its toggling between. Just so in the future if we have like 10 more functions to modify the 3d model scene this will be more clear what its doing

rknizzle reviewed (changes requested):
The switch toggle between views is working really really good now! And yea you already mentioned the couple things you need to finish up now in your commit messages.
- removing state no longer used
- getting rid of the pause/flash when you edit the threshold value

Nice detailed commit messages now too!

ErikAmerico committed:
Remove setOriginalFileNameFor3dModel

- this is no longer used in 3dModelViewer

ErikAmerico committed:
Use if-else statement instead of two if statements

ErikAmerico committed:
remove line number from comment

ErikAmerico committed:
Refactor modelViewer toggle logic

- use isChecked state to control the toggle button
  so the toggle always reflects the model being shown

- Before this change, when viewing a ply model,
  then switching to a different threshold value,
  the new ply model would be shown but the toggle
  button would be in the gray position.

ErikAmerico committed:
remove console.logs

ErikAmerico committed:
Create clearOriginalMesh Function

- clears the original mesh from the scene when
  a new part is selected

- I'm not sure why this wasn't necessary before,

- you'll see it used in the next commit

ErikAmerico committed:
Seperate the loading of originalFileFor3dModel
into its own useEffect hook

- the originalFileFor3dModel loading code only runs
  when a brand new file is selected or uploaded

- the fileFor3dModel loading code only runs when
  the when their is a file detected that doesn't
  match the originalFileFor3dModel. aka a ply file

- when manipulating the threshold input
  the original gray model isn't reloaded, the
  mesh colors get switched back to the original
  gray color

ErikAmerico committed:
Remove confusing current vs previous logic

- Now that theres a sepereate useEffect checking for
  a new originalFileFor3dModel, we can remove all
  of those variables and comparisons that were
  determining if we were working with a brand new
  file or not.

- In the first useEffect, when a new file is viewed,
  I am just sending `true` to loadModel() in
  setUpThreeScene, which will reposition the new
  model in the scene.

- In the second useEffect, I send `false`, since
  we are looking at either a certain ply file of
  same originalFileFor3dModel, or the
  originalFileFor3dModel itself.

ErikAmerico committed:
Rename: toggleMesh

-> toggleBetweenOriginalAndPlyMesh

ErikAmerico commented:
Also when adjusting the threshold value, The gray mesh gets displayed instead of reloading the original file. 

There is still the issue of when you type in a value that already has results, It loads in the ply file and you get that black flash. 

I know we talked about it probably not getting fixed. But I feel like after working tonight, I will be able to figure it out!

rknizzle commented:
Nice! Looks good merging this in
